tools/shared.py

```diff
@@ -2925,10 +2926,30 @@ def get_binaryen_feature_flags():
     ret += Settings.BINARYEN_FEATURES
     return ret
 
+  @staticmethod
+  def check_binaryen(bindir):
+    finalize = os.path.join(bindir, 'wasm-emscripten-finalize')
+    if not os.path.exists(finalize):
+      exit_with_error('binaryen executable not found (%s). Please check your binaryen installation' % finalize)
+    try:
+      output = run_process([finalize, '--version'], stdout=PIPE).stdout
+    except subprocess.CalledProcessError:
+      exit_with_error('error running binaryen executable (%s). Please check your binaryen installation' % finalize)
+    version = output.split()[2]
+    version = int(version)
+    # Allow the expected version or the following one in order avoid needing to update both
+    # emscripten and binaryen in lock step in emscripten-releases.
```

Comment:
doesn't this defeat most of the benefit of the check? +-1 is the most likely version mismatch I'd guess :(

Reply:
For consistency, if we want this +-1 approach, we should also do it with LLVM? But again, I don't think I like that.

Reply:
I'm more worried about people who install binaryen and emscripten in completely different ways, such as the homebrew of freebsd people.  They tend to not be off-by-one, but much more radical.   And when off by ones happen is normally that binaryen is too old I would have thought.  So I'm ok with it here.

We can see about generalize this mechanism and using shared code as a followup.  For example I'd like to make the llvm and java checks also happen on first use, and also be they only happen once per execution too.

For now I like this change though :)

Reply:
It was the FreeBSD issue that originally inspired this change: https://github.com/emscripten-core/emscripten/issues/10148